nixos/jupyter: init service

aborsu opened the pull request:
###### Motivation for this change
I believe that having a configurable service for jupyter with a template to write new kernels is a wanted feature.
cfr
https://www.reddit.com/r/NixOS/comments/7oujo4/getting_all_the_jupyter_kernels/
https://github.com/NixOS/nixpkgs/issues/10713

###### Things done
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This pull requests brings a jupyter service to nixos.
It contains a default bare python3 kernel and a literal example on how to define multiple kernels.

aborsu commented:
@FRidh @ixxie Any comments?

FRidh commented:
Related issue: https://github.com/NixOS/nixpkgs/issues/10713

FRidh commented on nixos/modules/services/development/jupyter.nix:
```diff
@@ -0,0 +1,236 @@
+      example = "users";
+    };
+
+    password = mkOption {
```
I think it is better to keep this out of the store. Maybe you can point to a path and read that during runtime?

aborsu replied:
I made a compromise on that, that I think should be agreeable to all.

FRidh commented on nixos/modules/services/development/jupyter.nix:
```diff
@@ -0,0 +1,236 @@
+            type = types.package;
+            default = (pkgs.python36.withPackages (
+              pythonPackages: with pythonPackages; [
+                notebook
```
`ipykernel` should be sufficient as the `notebook` should already be offered by this service

aborsu replied:
Indeed, good catch

FRidh commented on nixos/modules/services/development/jupyter.nix:
```diff
@@ -0,0 +1,236 @@
+          "python3" = {
+            name = "Python 3 for Machine Learning";
+            env = (pkgs.python36.withPackages (pythonPackages: with pythonPackages; [
+              notebook
```
same here

FRidh commented:
I think the logic of creating a `jupyter notebook` with a set of kernels can be factored out of the module and into a function in Nixpkgs.

Commit:
jupyter: init standalone package

Commit:
nixos/jupyter service

aborsu commented:
@FRidh I agree, but it was a bit more work.
I think it's pretty good for now. 
We could provide a few additional kernels but I think this should be done in other PR.

aborsu commented:
@FRidh @globin 
I've made a seperate package with the kernel logic, any feedback?
Let me know if you want me to change anything.
I've tested it on my server but so far only with Python kernels.
It would be great to have a Julia and R kernel so we could have the Ju Pyt R but i'm not very good in those two languages.

FRidh commented on pkgs/applications/editors/jupyter/default.nix:
```diff
@@ -0,0 +1,21 @@
+, lib
+, pkgs
+, python3
+, python3Packages 
```
`python3` or `python3Packages`

FRidh commented on pkgs/applications/editors/jupyter/default.nix:
```diff
@@ -0,0 +1,21 @@
+{ stdenv
+, lib
+, pkgs
```
no `pkgs`

FRidh commented on pkgs/applications/editors/jupyter/default.nix:
```diff
@@ -0,0 +1,21 @@
+python3Packages.buildPythonApplication rec {
+  inherit (python3Packages.notebook) pname version name src LC_ALL buildInputs propagatedBuildInputs checkPhase doCheck meta;
+
+  makeWrapperArgs = ["--set JUPYTER_PATH ${jupyterPath}"];
```
```
let
 ...
in with python3.pkgs; toPythonModule (notebook.override {
  makeWrapperArgs = ["--set JUPYTER_PATH ${jupyterPath}"];
});
```

`toPythonModule (buildPythonPackage {...})` is basically the same as `buildPythonApplication`. In this case I would go for the former, because the current approach requires keeping up with required attributes.

FRidh commented on pkgs/development/misc/jupyterKernels/default.nix:
```diff
@@ -0,0 +1,66 @@
+{pkgs, python3, lib, kernelDefinitions ? null }:
```
No `pkgs`

FRidh commented on pkgs/development/misc/jupyterKernels/default.nix:
```diff
@@ -0,0 +1,66 @@
+{pkgs, python3, lib, kernelDefinitions ? null }:
+
```
let
  defaultKernelDefinitions = {
    ...
  };
in {python3, lib, kernelDefinitions ? defaultKernelDefinitions }:

FRidh commented on pkgs/development/misc/jupyterKernels/default.nix:
```diff
@@ -0,0 +1,66 @@
+let
+  kernels = if kernelDefinitions != null
+    then kernelDefinitions
+    else {
```
this is then `defaultKernelDefinitions`

FRidh commented on pkgs/development/misc/jupyterKernels/default.nix:
```diff
@@ -0,0 +1,66 @@
+          "{connection_file}"
+        ];
+        language = "python";
+        logo32 = "${env}/lib/python3.6/site-packages/ipykernel/resources/logo-32x32.png";
```
`env.sitePackages`

FRidh reviewed (changes requested):
I think it is good to have a separate derivation for the kernel config, and having a new top level item for Jupyter. However, instead of making it a derivation that needs to be overriden, I suggest making `jupyter` a function:
```
kernelDefinitions: { the, other, args }: derivation
```

Commit:
jupyterKernel: init no version

This is a meta package for creating a jupyter notebook
kernel definition folder. It must be used in conjunction
with jupyter or jupyterhub to make different environements
availables.

aborsu commented:
@FRidh 
I didn't change everything exactly like you asked.

I'm not convinced making the top package jupyter a function is such a great idea.
The error message that I got when running `nix-env -i jupyter` or trying to add jupyter in my `environment.systemPackages` without giving it a kernelDefinition where unhelpful and I think will lead to a lot of head scratching by first time users. 
```
error: The option value `environment.systemPackages.[definition 1-entry 1]' in `...configuration.nix' is not of type `package'.
```
Even if I give a default argument, the function still needs to be called.

It seems to me that override is a standard way of providing customizability to packages and will avoid a lot of pain while imo not reducing understandability or ease of use.

As always I'm ready to discuss this further.

aborsu commented:
@FRidh ping?

aborsu commented on pkgs/development/misc/jupyterKernels/default.nix:
```diff
@@ -0,0 +1,76 @@
+in
+{
+  defaultKernelDefinition = defaultKernelDefinition;
+  create = { kernelDefinitions ?  defaultKernelDefinition }:
```
@FRidh I'm not certain this is better.
Now I'm exporting a set instead of exporting a function so I can re-use the default value in all the packages that might use it.

aborsu commented on pkgs/applications/editors/jupyter/default.nix:
```diff
@@ -0,0 +1,16 @@
+{ python3
+, jupyterKernels
+, kernelDefinitions ? jupyterKernels.defaultKernelDefinition
```
This is neat.

aborsu commented on nixos/modules/services/development/jupyter/default.nix:
```diff
@@ -0,0 +1,179 @@
+
+
+  kernels = (pkgs.jupyterKernels.create  {
+    kernelDefinitions = if cfg.kernels != null
```
I've basically had to move the null check here now.
I though of passing the `pkgs.jupyterKernels.defaultKernelConfig` as the default value for `jupyter.kernels`, but the it complains that it contains a reference to a derivation. (I can understand that having to build a derivation to create the doc is not the greatest).

aborsu commented:
@FRidh opinion?

schneefux reviewed (approved):
compiles and runs basic examples

dustinlacewell-wk commented:
Mind providing an example of how to get Jupyter installed with these changes?

schneefux commented:
Here is my configuration (the password is "test", change it):
```nix
  services.jupyter = {
    enable = true;
    port = 8123;
    notebookDir = "/var/www/jupyter";
    password = "'sha1:1b961dc713fb:88483270a63e57d18d43cf337e629539de1436ba'";
  };
```

schneefux commented on pkgs/development/misc/jupyterKernels/default.nix:
```diff
@@ -0,0 +1,76 @@
+      env = (pkgs.python3.withPackages (pythonPackages: with pythonPackages; [
+              ipykernel
+              pandas
+              scikitlearn
```
Are these Jupyter dependencies or just nice to have packages?

aborsu replied:
ipykernel is needed, the others are just basic packages used for data science.
I was thinking of getting something inline with the jupyter dockerimages.
Base python, base data-science, data-science + spark

FRidh replied:
Keep the default as minimal as possible.

teto commented:
Jupyter refuses to start when the noteBookdir does't exist, the folder existence could be check during the rebuild instead. I would like to use jupyter qtconsole with a sagemath kernel.
I am not sure how a kernel looks like, I believe it's a json file /nix/store/5i230cg5mixkdg6ah422fw783ahb2d27-sage-8.1/share/jupyter/kernels/sagemath/kernel.json ?

The definition of the  kernel is not crystal clear for a jupyter noob like me (do we really need to specify icons xD ?):
 ```
{
          python3 = let
            env = (pkgs.python3.withPackages (pythonPackages: with pythonPackages; [
                    ipykernel
                    pandas
                    scikitlearn
                  ]));
          in {
            displayName = "Python 3 for machine learning";
            argv = [
              "$ {env.interpreter}"
              "-m"
              "ipykernel_launcher"
              "-f"
              "{connection_file}"
            ];
            language = "python";
            logo32 = "$ {env.sitePackages}/ipykernel/resources/logo-32x32.png";
            logo64 = "$ {env.sitePackages}/ipykernel/resources/logo-64x64.png";
          };
```
Thanks for the PR

teto commented:
I rebased (since the current PR  has a network manager related problem) and managed to get a sagemath kernel working. I wonder if instead we could not change `displayName` by `display_name` as some packages already write their kernelspec to a json file (for instance "sage"), so we just have to read the json file (had to concat the missing "language" attribute though but that can be fixed in sage)
```
    kernels= {

      sage = let 
        kernel= builtins.fromJSON (builtins.readFile (pkgs.sage + "/share/jupyter/kernels/sagemath/kernel.json")) // {
          language="python";
        };
          env = (pkgs.python3.withPackages (pythonPackages: with pythonPackages; [
                    ipykernel
                    pandas
                    scikitlearn
                  ]));
        in 
        kernel;
```

teto commented:
I added myself in the jupyter group yet I could not read /var/lib/jupyter as it belonged to root (I can't remember if this is my fault or a nix problem when creating home).
Also I believe there might be a need to wrap jupyter as
```
nix-shell -p python3Packages.qtconsole
jupyter qtconsole --existing <json>

```
won't work http://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Connecting%20with%20the%20Qt%20Console.html
I have to chown/chmod /var/lib/jupyter and use the fullpath to the running instance
jupyter qtconsole --existing /var/lib/jupyter/.local/share/jupyter/runtime/kernel-e881c4f5-8115-406d-838a-888907eeb5b4.json

aborsu commented:
I'm sorry it seems that trying to make a PR from the 18.03 release was not such a good idea... I've rebased my work on unstable.

ixxie commented:
Cool, I will check it out when I get some time.



I plan to draft a sort of RFC issue for Nixpkgs where we could coordinate data science infrastructure work. I would like to provide an overview or stuff related to the Jupyter ecosystem, Python data science ecosystem and Haskell data science ecosystem; perhaps other stuff too, but scoping this is a little tricky, so I think that is an okay start.



I plan to make a script to query nix search (which can return JSONs now) for a package list. datahaskell.org provides a nice overview of interesting Haskell stuff but I would still need to find a decent Python list.



Any thoughts?




 * * * * *

Matan Bendix Shenhav
@ixxie on GitHub &amp; Freenode

+358 (0)45 6 135 315






---- On Sat, 14 Apr 2018 15:09:06 +0300 Augustin Borsu &lt;notifications@github.com&gt; wrote ----




I'm sorry it seems that trying to make a PR from the 18.03 release was not such a good idea... I've rebased my work on unstable.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

aborsu commented:
@ixxie for python, using the packages used by the jupyter docker images would be a good start https://github.com/jupyter/docker-stacks/tree/master/datascience-notebook

ixxie commented:
@aborsu thanks for the tip

Sorry everybody about that offtopic-ish post; I didn't realize I could reply to an email in my mail client and it would post to GitHub!

phile314-fh commented:
I have been using this for the last few weeks and it works nicely :+1:

phile314-fh commented:
I have set ``notebookDir`` to ``/var/www/jupyter`` and it works fine for me. I don't remember though if I created the notebookDir by hand and set the permissions accordingly or if the jupyter service did it automatically.

ixxie commented:
@phile314-fh  - looks like you did it manually. With @cleverca22's help I managed to resolve this by adding:

        serviceConfig.PermissionsStartOnly = true

and
        
        preStart = ''
          mkdir -p ${cfg.notebookDir}
          chown ${cfg.user}:${cfg.group} ${cfg.notebookDir}
        '';

to `systemd.services.jupyter`.

aborsu commented:
Hello, sorry, I've been a bit busy.
@FRidh any comments before I start hacking this again?

FRidh commented on nixos/modules/services/development/jupyter/default.nix:
```diff
@@ -0,0 +1,179 @@
+
+        wantedBy = [ "multi-user.target" ];
+
+        path = [ pkgs.bash ]; # needed for sh in cell magic to work
```
Have you considered patching `notebook` to hardcode `bash`?

FRidh commented on pkgs/applications/editors/jupyter/default.nix:
```diff
@@ -0,0 +1,16 @@
+
+let
+
+  jupyterPath = (jupyterKernels.create  { kernelDefinitions =  kernelDefinitions; });
```
inherit / space

FRidh commented on nixos/modules/services/development/jupyter/kernel-options.nix:
```diff
@@ -0,0 +1,60 @@
+    argv = mkOption {
+      type = types.listOf types.str;
+      example = [
+        "{customEnv}/bin/python"
```
${customEnv.interpreter}

FRidh commented on nixos/modules/services/development/jupyter/kernel-options.nix:
```diff
@@ -0,0 +1,60 @@
+    logo32 = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      example = "{env}/lib/python3.6/site-packages/ipykernel/resources/logo-32x32.png";
```
${env.sitePackages}

FRidh commented on nixos/modules/services/development/jupyter/kernel-options.nix:
```diff
@@ -0,0 +1,60 @@
+    logo64 = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      example = "{env}/lib/python3.6/site-packages/ipykernel/resources/logo-64x64.png";
```
${env.sitePackages}

FRidh commented on pkgs/applications/editors/jupyter/default.nix:
```diff
@@ -0,0 +1,16 @@
+in
+
+with python3.pkgs; toPythonModule (
+  notebook.overrideAttrs(oldAttrs: {
```
overridePythonAttrs

FRidh commented on pkgs/development/misc/jupyterKernels/default.nix:
```diff
@@ -0,0 +1,76 @@
+{ lib, pkgs, stdenv}:
```
no `pkgs`, but `python3`

FRidh commented on pkgs/development/misc/jupyterKernels/default.nix:
```diff
@@ -0,0 +1,76 @@
+    meta = {
+      description = "Wrapper to create jupyter notebook kernel definitions";
+      homepage = http://jupyter.org/;
+      # NIXOS license as this is a nixos meta package.
```
No need

FRidh commented on pkgs/development/misc/jupyterKernels/default.nix:
```diff
@@ -0,0 +1,76 @@
+
+    src = "/dev/null";
+
+    unpackCmd ="mkdir jupyter_kernels";
```
space

FRidh commented on pkgs/development/misc/jupyterKernels/default.nix:
```diff
@@ -0,0 +1,76 @@
+
+    unpackCmd ="mkdir jupyter_kernels";
+
+    installPhase =  ''
```
space

FRidh reviewed (commented):
I don't really have the time to review this any further, so please ask also other maintainers if they can help it forward / get it in.

dotlambda commented:
Not until Sunday or even next week, sadly.

dotlambda commented:
@ixxie Please fix the things FRidh mentioned. I'll then take a look. However, I've never worked with Jupyter on NixOS so I might need some time to read through it.

ixxie commented:
@dotlambda well this is @aborsu's PR really; I am working on the Jupyterlab module (no PR for that yet).

teto commented:
Really looking forward to this. Keep up the good work @aborsu :D

FRidh commented:
I've made some minor changes and pushed to master 4d3ce5ca36a145748a4938cb0127c115d8fbacf0.

teto commented:
I'll try to give some feedback on the feature, has anyone tried it with an haskell kernel already ?
The jupyter service might deserve a paragraph in the manual (at least a mention in changelog)

timokau commented:
In https://github.com/NixOS/nixpkgs/pull/51030 I add jupyter support for sage. I've added it so that `sage -n jupyter` can start a jupyter notebook with the sagemath kernel. I'm not sure how that would integrate with this service, but I've used the kernel spec definition from here and made it available as `sage.kernelspec`. So if someone wants to integrate that into the jupyter service, that'd be nice.
